distml/strategy/ps_strategy.py

```diff
@@ -0,0 +1,702 @@
+            self.worker_group.actors[0].params_distribution.remote())
+        assignments = [0 for _ in parameter_distribution]
+        loads = [0 for _ in range(self.num_ps)]
+        for i, var_size in enumerate(parameter_distribution):
+            min_ps_index = loads.index(min(loads))
+            loads[min_ps_index] += var_size
+            assignments[i] = min_ps_index
+        print("Load of each ps {}".format(loads))
+        self.assignments = assignments
+
+    def train(self, num_steps=None):
+        # TODO (Hao): add fault tolerance using `max_retries`.
+        steps = num_steps if num_steps \
+            else self.worker_group.get_data_loader_len()
+
+        # TODO(HUI): Record server rank instead of using num_ps.
+        # TODO(Hao): this call should be hidden inside Replica.
+        # train one epoch
+        self.worker_group.make_iterator()
+        for idx in range(steps):
+            with self._collector.record("train"):
+                metrics = self.train_batch()
+            logger.info("Step: {}/{}".format(idx, steps))
+        return metrics
+
+    def validate(self, num_steps=None):
+        steps = num_steps if num_steps \
+            else self.worker_group.get_data_loader_len(training=False)
+        self.worker_group.make_iterator(training=False)
+
+        # TODO(HUI): Construct a better tool to save validate results.
+        for idx in range(steps):
+            batch_metrics = self.worker_group.validate_batch()
+        # Validate results should be the same in all workers
+        return batch_metrics
+
+    def train_batch(self):
+        loss_vals = []
+        rets = []
+        metrics = {}
+
+        for worker_idx, worker in enumerate(self.worker_group.actors):
+            for server_idx, server in enumerate(self.server_group.actors):
+                # every server sends its shard to the worker
+                server.send_params.remote(worker_idx)
+            # the worker receives shards from ps, compute loss, gradients
+            # and sends these gradients to every server
+            loss_val = worker.compute.remote()
+            loss_vals.append(loss_val)
+
+        for worker_idx, worker in enumerate(self.worker_group.actors):
+            for server in self.server_group.actors:
+                rets.append(server.update.remote(worker_idx))
+
+        loss_vals = ray.get(loss_vals)
+        ray.get(rets)
+        train_loss_list = [d["train_loss"] for d in loss_vals]
+        metrics["train_loss"] = np.mean(train_loss_list)
+        return metrics
+
+
+class PS(object):
+    def __init__(self, training_operator_cls, operator_config):
+        self.training_operator_cls = training_operator_cls
+        self.operator_config = operator_config
+
+        self.grad_counts = None
+        self.params = dict()
+
+    def setup_operator(self):
+        # figure out the signature of training_operator_cls later.
+        self.training_operator = self.training_operator_cls(
```

Comment:
I am not sure whether we should setup the whole operator on the server side? One drawback is that this will take a lot of GPU memory?